### DISCUSSION TITLE
Unique version numbers for every working copy / every commit

### DISCUSSION THREAD

**Discussion by bleykauf** [pymeasure/pymeasure#566] [Ideas]
What do you think about including automatic version numbering like [versioneer](https://github.com/python-versioneer/python-versioneer) to pymeausure? You would have unique version numbers for every commit even in forks and "dirty" repo states. 

For my personal data analysis I use versioneer and print the `.__version__` dunder method after importing to document a specific state of the software (possibly containing bugs) in my notebooks to be sure to be able to later replicate the results. Right now if I use pymeausure, whether installed via `pip install pymeasure` or by cloning the repo and calling `pip install -e .` (for editable mode), I would always get the version `0.9.0` even though in the current master branch there have been lots of changes. For a new release this number would have to be consistently changed in several places.

Versioneer would use the git repo data, i.e. tag names (for the "proper" version `.0.9.2.` etc and the commits by creating versions like `v0.9.2+4.g10784c97`. 

It also does play very nice with automatically deploying new (beta?) versions to pypi.

**Comment by bilderbuchi** [comment]
Hm, does `git describe` not help you determine the version?

I haven't used this yet, and just skimmed the Readme, but it seems pretty invasive - placing some files around the code base, and replacing setuptools stuff.
In some personal projects, I use a `__version__.py` file or somesuch as a single source for the version string, so just one place to update. 

This tool looks like it could be useful, though, although considering https://github.com/python-versioneer/python-versioneer/issues/193, a our use of conda, and a slowly modernizing Python packaging approach (causing things like [this](https://github.com/python-versioneer/python-versioneer/issues/275)), maybe rather [this version](https://anaconda.org/conda-forge/versioneer-518). Not a fan of vendored dependencies (creating manual update burden), typically.

At our release cadence, though, *cough*, the effort of updating a handful of strings is a negligible waste :D

@cjermain what do you think?

**Comment by bleykauf** [comment]
Sure, I could in principle use git to get info about the repos state. I was thinking more about storing the specific version number in the stored data, cp. https://github.com/pymeasure/pymeasure/pull/503. If you use pymeasure from an Jupyter notebook, it is also way easier to print `pymeasure.__version__` than invoking git to save this info in the file.

I think using single file to store the version would already be helpful I think. Versioneer actually uses a `_version.py` and the setuptools simply link to that file (as you would if you simply stored a string with a version number in there). Other than that it places a `versioneer.py` file in the projects root.

Thank you for pointing out some persisiting issues with versioneer and versioneer-518, I was not aware of them.

**Reply by msmttchr** [reply]
Complexity can be evaluated having a look at PR #574

**Reply by bilderbuchi** [reply]
should be resolved by #613
